deepinv/__main__.py

```diff
@@ -0,0 +1,284 @@
+                backbonet_net_cls: type = getattr(dinv.models, backbonet_net_cls_name)
+                if not issubclass(backbonet_net_cls, dinv.models.Denoiser):
+                    raise ValueError(
+                        f"Could not find denoiser class '{backbonet_net_cls_name}' in 'deepinv.models'."
+                    )
+
+                backbonet_net_args: list = backbonet_net_config.pop("args", [])
+                if not isinstance(backbonet_net_args, list):
+                    raise ValueError(
+                        f"The 'model.kwargs.backbonet_net.args' entry must be a list, got: {type(backbonet_net_args)}."
+                    )
+
+                backbonet_net_kwargs: dict = backbonet_net_config.pop("kwargs", {})
+                if not isinstance(backbonet_net_kwargs, dict):
+                    raise ValueError(
+                        f"The 'model.kwargs.backbonet_net.kwargs' entry must be a dictionary, got: {type(backbonet_net_kwargs)}."
+                    )
+
+                if backbonet_net_config:
+                    raise ValueError(
+                        f"The 'model.kwargs.backbonet_net' entry contains unexpected keys: {list(backbonet_net_config.keys())}."
+                    )
+
+                backbone_net: dinv.models.Denoiser = backbonet_net_cls(
+                    *backbonet_net_args, **backbonet_net_kwargs
+                )
+                if "backbone_net" in model_kwargs:
+                    raise ValueError(
+                        "The 'model.kwargs' entry contains a 'backbonet_net' key, which is reserved."
+                    )
+                model_kwargs["backbone_net"] = backbone_net
+
+            if model_config:
+                raise ValueError(
+                    f"The 'model' entry contains unexpected keys: {list(model_config.keys())}."
+                )
+
+            model: dinv.models.Reconstructor = model_cls(*model_args, **model_kwargs)
+
+        except KeyError as e:
+            (key,) = e.args
+            raise ValueError(f"The mandatory entry for key '{key}' is missing.")
+
+        if config:
+            raise ValueError(
+                f"The configuration dictionary contains unexpected entries. Keys: {list(config.keys())}."
+            )
+
+        return cls._ParsedConfig(
+            dataset=dataset,
+            physics=physics,
+            model=model,
+        )
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Train a model using DeepInverse.")
+
+    subparsers = parser.add_subparsers(dest="command")
+    train_parser = subparsers.add_parser("train", help="Train a model.")
+    train_parser.add_argument(
+        "--config",
+        type=str,
+        required=True,
+        help="Path to the YAML configuration file.",
+    )
+
+    args: argparse.Namespace = parser.parse_args()
+
+    if args.command == "train":
+        trainer = CommandLineTrainer()
```

Comment:
Maybe we could have a `cli` submodule which contains all the orchestrators e.g. `CommandLineTrainer`, `CommandLineOptim` etc. to keep this `__main__` clean. And also keep this argparser clean, and the individual files in `cli` can contain the task-specific args